### PR TITLE
ENH: use linux fallocate to reserve diskspace in array.tofile

### PR DESCRIPTION
**Pull request by juliantaylor**
fallocate allows the filesystem to make smarter decisions about space
allocation and gives a fast failure path for insufficient space.
This is very important for filesystems that suffer a lot from
fragmentation like btrfs.
Restricted to linux only as that is the only system I know the behavior
of. Other systems might also have this system call but we don't want to
accidentally trigger explicit zeroing behavior as e.g. posix_fallocate
would when there is no support for a real fallocate.

### CODE REVIEW

File: numpy/core/tests/test_multiarray.py [hunk 1]
```diff
@@ -3377,6 +3378,18 @@ def fail(*args, **kwargs):
         y = np.fromfile(self.filename, dtype=self.dtype)
         assert_array_equal(y, self.x.flat)
 
+    def test_largish_file(self):
+        # check the fallocate path on files > 16MB
+        d = np.zeros(4 * 1024 ** 2)
```
Comment (mhvk) [inline numpy/core/tests/test_multiarray.py]: This won't actually trigger `fallocate`, since it is less than `16 * 1024 * 1024`. Is that intentional?

Comment (juliantaylor) [inline numpy/core/tests/test_multiarray.py]: the default dtype is 8byte floats so its an 32mb array.

Comment (mhvk) [inline numpy/core/tests/test_multiarray.py]: Duh.  Sorry for the noise.

**Comment by charris** [comment]
@juliantaylor Could you add a note under `Improvements` in the 1.11.0 release notes?

**Comment by charris** [comment]
@juliantaylor Needs a rebase.

**Comment by charris** [comment]
Thanks Julian.

**Pull request closed without merging**